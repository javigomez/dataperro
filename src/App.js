import React, {useState} from 'react';
import styled from 'styled-components'

export const Widget = styled.li`
background: #e2dfdd;
margin: 10px;
width: 25vw;
height: 25vh;
list-style: none;
border-radius: 8px;
border: 2px solid #a4a3a3;
`
const FullScreenWidget = styled.div`
background: #e2dfdd;
margin: 60px;
list-style: none;
border-radius: 8px;
border: 2px solid #a4a3a3;
position: absolute;
top: 0;
bottom: 0;
left: 0;
right: 0;
display: flex;
align-items: center;
justify-content: center;
`

const Wrapper = styled.ul`
  display: flex;
`

export const TimelapseWidget = (props) => {
  if(!props.data) return 'no data'

  return props.data.map((number, index) => {
    return (<div key={index}>{number}</div>)
  }) 
}

export const NumberWidget = (props) => {
  if(!props.data) return 'no data'

  return (JSON.stringify(props.data))  
}


export const WidgetFactory = ({data, kind}) => {
  switch (kind) {
    case 'number':
      return <div><NumberWidget data={data}/></div>
    case 'timelapse':
      return <TimelapseWidget data={data} />
    default:
      return null
  }
}


function App(props) {
  
  const [selectedWidget, setSelectedWidget] = useState(props.initialData.selectedWidget)

  const widget = props.initialData.widgets.find(widget => widget.id === selectedWidget)

  const widgetList = props.initialData.widgets.map(widget => {
    function handleClick () {
      setSelectedWidget(widget.id)
    }

    return (<Widget key={widget.id} onClick={handleClick}><WidgetFactory data={widget.data} kind={widget.kind}/></Widget>)
  } ) 

  return (
    <div className="App">
      <Wrapper>
        {widgetList}
      </Wrapper>
      {widget &&  <FullScreenWidget><WidgetFactory data={widget.data} kind={widget.kind}/></FullScreenWidget>}
    </div>
  );
}

export default App;
