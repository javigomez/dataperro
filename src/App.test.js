import React from 'react'
import ReactDOM from 'react-dom'
import App, {WidgetFactory, NumberWidget} from './App'
import TestRenderer from 'react-test-renderer'
import expectExport from 'expect';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App initialData={{ widgets: []}} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('WidgetFactory', () => {
  it('should render number widget when kind is number', () => {
    const testInstance = TestRenderer.create(<WidgetFactory data='2' kind='number' />).root
    //expect(testInstance).toMatchSnapshot()
    expect(testInstance.findByType(NumberWidget)).toBe(true)
  })
})
