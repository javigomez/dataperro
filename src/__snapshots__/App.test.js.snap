// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetFactory should render number widget when kind is number 1`] = `
<div>
  "2"
</div>
`;
